content/posts/outbox.md

```diff
@@ -0,0 +1,74 @@
+
+Let's take a look at the code to understand better all the moving parts. First, we need to load the state of our business component. 
+
+```
+var stateContainer = await persister.LoadByCorrelationId(correlationId).ConfigureAwait(false) ?? new StateContainer { Id = correlationId };
+
+TransportOperation[] outgoingMessages;
+if (!stateContainer.OutboxState.ContainsKey(messageId))
+{
+```
+
+As you can see, the state container is extended with a data structure that holds the message processing state -- `OutboxState`. That structure is a map that associates ID of an incoming message with a collection of outgoing messages that were produced during processing. When a message handler receives a message, it loads the state and inspects the _Outbox_ to check if it contains an entry for the ID of the incoming message. If there is no corresponding entry, it means that a given message has not been processed yet.
+
+```
+var state = (T)stateContainer.State ?? new T();
+var (newState, pendingTransportOperations) = await handlerCallback(state, context);
+```
+
+In that case we invoke the business logic that returns a tuple containing a new value of the state and a collection of `pendingTransportOperations` -- messages that were generated and are to be dispatched. Now we need to take both these pieces, stick them into the container and make sure they are stored durably in a *single atomic operation*.
+
+```
+outgoingMessages = pendingTransportOperations.Operations.Serialize();
+
+stateContainer.State = newState;
+stateContainer.OutboxState[messageId] = new OutboxState
+{
+    OutgoingMessages = outgoingMessages
+};
+
+await persister.Persist(stateContainer);
+```
+
+Let's go back and look at the other branch of the `if` statement. What if the condition
+
+```
+if (!stateContainer.OutboxState.ContainsKey(messageId))
+{
+```
+
+is false. What if the _Outbox_ state already contains an entry for a given ID? That means we have already processed a another copy of the incoming message. There are two possible reasons for this. The message may simply be a duplicate. Another possibility is that there exists a single copy of that message but the first time we attempted to process it, we failed to dispatch the outgoing messages and the incoming message has been returned to the queue. 
+
+```
+outgoingMessages = stateContainer.OutboxState[messageId].OutgoingMessages;
+```
+
+If a message is indeed a duplicate, the collection of outgoing messages is `null` and we don't need to do anything. You may be wondering now what happens if two copies of a message come in at exactly the same time. It may be that one thread manages to persist the _Outbox_ state before the other begins the processing. In that case the second thread will see the outgoing messages as not `null` even though the message is a duplicate. As a result, the outgoing message will be dispatched two times, however this is not a problem as the receivers of these messages will be able to de-duplicate them.
```

Comment:
It might be not clear why in such cases the outgoing message collection is `null`.